wordpress-mobile/WordPress-FluxC-Android
# Issue/1335 v4 visitor stats changes

Fixes #1335. This PR adds logic to get the visitor stats data for the new date ranges as defined in [here](https://github.com/woocommerce/woocommerce-android/issues/1199). We would use the existing API endpoint `/rest/v1.1/sites/{{site_id}}/stats/visits`. But as described [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/1335), the `quantity`, `unit` and `date` fields would need to be modified to satisfy the new v4 stats UI.

Since we will be supporting both stats UI for the foreseeable future, I thought it made sense to implement the v4 changes for the visitor stats separately for better code readability and maintenance.

### Changes:
- Created a new table and model class - `WCNewVisitorStatsModel`. - The only difference between this model class and the `WcVisitorStatsModel` class is that we are storing the `StatsGranularity` to the `WCNewVisitorStatsModel` instead of `OrderStatsApiUnit`. This is because the current visitor stats implementation fetches the visitor data by `WEEKS` or `MONTHS`. In the new stats UI, we would need to fetch the visitor stats in `DAYS` for the current week or month. 
- Adds an action to `WCStatsAction`: `FETCH_NEW_VISITOR_STATS` and `FETCHED_NEW_VISITOR_STATS`  - for fetching visitor stats with a different `StatsGranularity` and `quantity` and handling visitor stats response. 
- Adds two payload classes to `WcStatsStore`: `FetchNewVisitorStatsPayload` and `FetchNewVisitorStatsResponsePayload`. 
- Adds a new method in `OrderStatsRestClient`: `fetchNewVisitorStats` to store the response data separately. This will be used by the new stats UI.
- Adds new buttons to the `WooRevenueStatsFragment` to test the visitor stats api for the current day, week, month and year and verify that the data is consistent with the date range.
- Added mocked tests to `MockedStack_WCStatsTest`
- Keeping in mind @nbradbury 's suggestion [here](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1325#pullrequestreview-268432017), created a new `StoreSelectorDialog` that displays only WooCommerce sites :)  

### Testing:
- Run all tests in `MockedStack_WCStatsTest`
- Run all tests in `WCStatsStoreTest`
- Run all tests in `WCVisitorStatsSqlUtilsTest`.
- Verify that the visitor stats fields are consistent with the `StatsGranularity` selected. Also note that the size of the visitors data map fetched from the local cache should match the `quantity` field:

| `TAB`  | `quantity` field | `date` field | `unit` field |
| ------------- | ---------------- | ---------------- | ---------------- |
| TODAY  | 1  | current date i.e. `2019-08-06` |  `OrderStatsApiUnit.DAY`  |
| This WEEK  | 3  | current date i.e. `2019-08-06` |  `OrderStatsApiUnit.DAY` (since we need the data in days i.e. for each day from the first of this week till the current date)  |
| This MONTH  | 6  | current date i.e. `2019-08-06` |  `OrderStatsApiUnit.DAY` (since we need the data in days i.e. for each day from the first of this month till the current date)  |
| This YEAR  | 8  | current date i.e. `2019-08-06` |  `OrderStatsApiUnit.MONTH` (since we need the data in months) - The API currently responds with 9 quantities, even though we have specified it as 8.  |

- Verify that the current implementation of visitor stats is working correctly, as a sanity check.
- Verify that upgrading to the changes in this PR does not lead to a crash.

### Screenshots
<img width="300" src="https://user-images.githubusercontent.com/22608780/62520226-c133d100-b84a-11e9-8654-3ea5b5e42276.png">
